docs/architecture/adr-060-abci++-phase-i.md

```diff
@@ -0,0 +1,217 @@
+   one of two buffers -- _valid_ and _invalid_.
+2. For transactions that DO NOT violate the nonce validation, they are included
+   in the _valid_ buffer.
+3. For transactions that DO violate the nonce validation, they are included in
+   the _invalid_ buffer.
+4. Continue this process until the desired number of valid transactions are
+   reaped or until the mempool is empty.
+5. Provide Tendermint the list of all transactions from the _valid_ buffer.
+6. Re-insert all transactions, from both buffers, back into mempool. This is to
+   ensure we do not discard transactions from the mempool in case `ProcessProposal`
+   fails.
+
+```go
+type PriorityMempool[T MempoolTx] struct {
+	queue   *PriorityQueue[MempoolTx]
+	senders map[string][]int64
+	// ...
+}
+```
+
+Previous discussions<sup>1</sup> have come to the agreement that Tendermint will
+perform a request to the application, via `PrepareProposalRequest`, with a certain
+amount of transactions reaped from Tendermint's local mempool. The exact amount
+of transactions reaped will be determined by a local operator configuration.
+This is referred to as the "one-shot approach" seen in discussions.
+
+When Tendermint reaps transactions from the local mempool and sends them to the
+application via `PrepareProposalRequest`, the application will have to evaluate
+the transactions. Specifically, it will need to inform Tendermint if it should
+reject and or include each transaction. Note, the application can even _replace_
+transactions entirely with other transactions.
+
+When evaluating transactions from `PrepareProposalRequest`, the application will
+ignore _all_ transactions sent to it in the request and instead reap up to
+`PrepareProposalRequest.max_tx_bytes` from it's own mempool. There is no need to
+execute the transactions for validity as they have already passed CheckTx.
+
+### `ProcessProposal`
+
+The `ProcessProposal` ABCI method is relatively straightforward. It is responsible
+for ensuring validity of the proposed block containing transactions that were
+selected from the `PrepareProposal` step. In order to check validity of the proposed
+block, we must iterate over the list of transactions and execute them using the
+same mode/execution strategy as `CheckTx`. In other words, we execute each transaction
+using the AnteHandler only -- no messages are executed. However, we cannot just
+execute `CheckTx` again, because `BaseApp` already has a modified `checkState` at
+this point. So when executing `ProcessProposal`, we create a similar branched
+state, `processProposalState`, off of `deliverState`. Using `processProposalState`
+we execute the AnteHandler for each transaction. Note, the `processProposalState`
+is never committed and is completely discarded after `ProcessProposal` completes.
+
+We will only populate the `Status` field of the `ProcessProposalResponse` with
+`ACCEPT` if ALL the transactions were accepted as valid, otherwise we will
+populate with `REJECT`.
```

Comment:
I'm assuming it's never possible for two different nodes to have different views of transaction validity in this scheme? Bringing it up because ABCI++ requires coherence between prepare and process proposal to prevent liveness issues. (see requirement 5 here: https://github.com/tendermint/tendermint/blob/master/spec/abci%2B%2B/abci%2B%2B_app_requirements.md).

Reply:
It should hold true as state across the Network is the same. As things like MEV begin being deployed the operators offering txs to be included in blocks will need to be more intelligent

Reply:
I realize my comment above is somewhat redundant w.r.t this one from @williambanfield's.